### PR TITLE
Bluetooth Audio Support

### PR DESCRIPTION
**Pull request by wansti**
# Bluetooth Audio Support

## Description

- [X] New feature (non-breaking change which adds functionality)

This adds basic functionality required for bluetooth audio, namely the bluez-alsa package and several audio codec libraries. An additional daemon named bluealsa will run in the background to handle the link between the bluetooth and audio subsystems. You can pair your bluetooth headsets in the Controllers & Bluetooth menu; paired devices will then appear in the System / Audio Device menu and connect upon selection. Supported codecs are SBC, AAC, APT-X and LDAC.

## Known issues:
- Does not gracefully handle disconnects. If you disconnect your headset without first switching back to the default audio device, you won't get any audio until you reboot.

- With some headsets, the bluetooth daemon occasionally crashes after exiting a game. This happens since the last stable release (20221114) for unknown reasons.
**Workaround**: Reboot (or turn off bluetooth and turn it back on again), then reconnect the headset.

- Does not work with the PCSX ReARMed32 Retroarch core for unknown reasons (no sound will play from this core).
**Workaround**: Use a different PSX core (e.g. PCSX ReARMed, SwanStation) when you want to play with a bluetooth headset.

- Headset volume can't be controlled with the console's volume buttons (they will still be mapped to the default card).
**Workaround**: Use the headset's hardware controls.

- Cannot test AAC codec due to lack of compatible hardware. If you have compatible headphones (e.g. Apple Airpods) please let me know whether they are working.

## How Has This Been Tested Locally?

- [X] Built for Anbernic RG353V
- [X] Verified existing functionality
- [X] Connected several pairs of bluetooth headphones and verified audio is working as expected
- [X] Tested with a variety of emulators
- [X] No noticeable impact on emulator performance
- [X] Headset test results: https://github.com/wansti/distribution/wiki#bluetooth-audio-test-matrix

**Test Configuration**:
* Build OS name and version: Ubuntu 20.04
* Docker (Y/N): Y
* JELOS Branch: dev
* Any additional information that may be useful:
  Change in set_audio causes asound.conf to be directly overwritten with /usr/config/asound.conf.RG353V on that device - is there a situation where that might be a problem?

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [N/A] I have made corresponding changes to the documentation

### CODE REVIEW

File: packages/audio/bluez-alsa/package.mk [hunk 1]
```diff
@@ -0,0 +1,42 @@
+# SPDX-License-Identifier: GPL-2.0-or-later
+# Copyright (C) 2009-2016 Stephan Raue (stephan@openelec.tv)
+# Copyright (C) 2016-present Team LibreELEC (https://libreelec.tv)
+# Copyright (C) 2022-present Fewtarius
```
Comment (fewtarius) [inline packages/audio/bluez-alsa/package.mk:4]: These aren't my packages, they probably shouldn't have my copyright applied to them.

Comment (wansti) [reply packages/audio/bluez-alsa/package.mk]: Fair point - fixed.

File: packages/jelos/sources/scripts/set-audio [hunk 2]
```diff
@@ -47,7 +54,7 @@ function set_audio_device() {
   then
     if [ "$(get_setting system.rg353v)" = "1" ]
     then
-      cp /usr/config/asound.conf.RG353V /storage/.config
```
Comment (wansti) [inline packages/jelos/sources/scripts/set-audio:50]: Is this safe to do?

File: packages/jelos/sources/scripts/set-audio [hunk 3]
```diff
@@ -47,7 +54,7 @@ function set_audio_device() {
   then
     if [ "$(get_setting system.rg353v)" = "1" ]
     then
```
Comment (wansti) [inline packages/jelos/sources/scripts/set-audio:56]: Is it safe to directly overwrite asound.conf on the RG353V, or will this cause issues in other places?

Comment (fewtarius) [reply packages/jelos/sources/scripts/set-audio:55]: Yep, as long as you also expect to restart the emulator or ES.  May also need to bounce the volume service.

File: packages/audio/fdk-aac/package.mk [hunk 4]
```diff
@@ -0,0 +1,20 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (C) 2022-present Marek Moeckel (wansti@discarded-ideas.org)
+
+PKG_NAME="fdk-aac"
+PKG_VERSION="2.0.2"
+PKG_SHA256="7812b4f0cf66acda0d0fe4302545339517e702af7674dd04e5fe22a5ade16a90"
+PKG_LICENSE="Apache2"
```
Comment (fewtarius) [inline packages/audio/fdk-aac/package.mk:7]: I'm concerned about the license on this, as it's not open source compatible.

https://github.com/mstorsjo/fdk-aac/blob/master/NOTICE

https://www.gnu.org/licenses/license-list.en.html#fdk

Comment (wansti) [reply packages/audio/fdk-aac/package.mk]: Updated as per our discussion on Discord.

**Comment by GavinDarkglider** [comment]
Hey, lakka dev here...... Just thought I should mention, that adding pulse audio to the build fixes the headset disconnect issues, and volume button issues...... Not saying this doesnt have merrit, just saying the end result is easier to achieve with less changes since the package is already in the tree, only the configs are missing, and you can steal that from lakka.

**Comment by wansti** [comment]
> Hey, lakka dev here...... Just thought I should mention, that adding pulse audio to the build fixes the headset disconnect issues, and volume button issues...... Not saying this doesnt have merrit, just saying the end result is easier to achieve with less changes since the package is already in the tree, only the configs are missing, and you can steal that from lakka. Pipewire would work too, but on some boards, it has issues last I tried it.

Thanks for the advice! Have you tested it? Are there any implications for performance? And where can I find those configs?

**Comment by GavinDarkglider** [comment]
Currently I am using it on 2 builds I maintain, the Nintendo Switch, and the Ayn Odin.....

Config package can be found here: https://github.com/libretro/Lakka-LibreELEC/tree/Lakka-v5.x/packages/audio/pulseaudio-configs

This should just be a copy paste into your tree, then add pulseaudio to the build you are using...... Basically, with this setup, alsa uses pulseaudio as pcm, so all audio that gets sent to alsa should get wrapped through pulseaudio, then back to the chosen device..... Also handles autoswitching when normal headphones/headset are plugged in. As for implications for performance, I have found in some cores in retroarch it is better to use the alsathread backend, over the pulseaudio backend..... either way pulse audio gets used, but the alsathread backend is better it seems.

**Comment by wansti** [comment]
Does this replace BlueALSA then?

**Comment by GavinDarkglider** [comment]
Originally all bluetooth audio was handled by pulseaudio, so it is more like blueALSA was created to replace the need for pulseaudio when using bluetooth devices. Now upstream developers are replacing alsa with pipewire, which does more than pulse, but also breaks some things that worked fine in pulseaudio...... As I said, do your own tests, but I think 1 added set of configs, and adding pulseaudio to your builds would be a hell of a lot easier to review than the 60 commits in this PR...... just my 2 cents...... And it comes with the added benifit of fixing all of the issues listed in the op of this PR.

**Pull request merged**